rest/rest.go

```diff
@@ -0,0 +1,348 @@
+	r.files[name] = value
+	return r
+}
+
+// Body sets body parameter.
+func (r *Request) Body(contentType string, value interface{}) *Request {
+	r.body = value
+	r.bodyContentType = contentType
+	return r
+}
+
+// Meta sets header result.
+func (r *Request) Meta(value map[string]string) *Request {
+	r.meta = value
+	return r
+}
+
+// Data sets body result.
+func (r *Request) Data(value interface{}) *Request {
+	r.data = value
+	return r
+}
+
+// Do executes the request.
+func (r *Request) Do(ctx context.Context) error {
+	if r.body != nil && (len(r.files) > 0 || len(r.forms) != 0) {
+		return fmt.Errorf("conflict body parameter in path %s", r.path)
+	}
+	path := r.endpoint
+	for _, v := range r.segments {
+		value := v.value
+		if v.name != "" {
+			p := r.paths[v.name]
+			if p == nil {
+				return fmt.Errorf("can't find path parameter %s in path %s", v.name, r.path)
+			}
+			value = fmt.Sprint(p)
+		}
+		path += value
+	}
+	urlVal := url.Values{}
+	for k, v := range r.queries {
+		urlVal.Add(k, fmt.Sprint(v))
+	}
+	if len(urlVal) > 0 {
+		path += "?" + urlVal.Encode()
+	}
+	contentType := r.bodyContentType
+	buf := bytes.NewBuffer(nil)
+	reader := io.Reader(buf)
+	if r.body != nil {
+		if body, ok := r.body.(io.Reader); ok {
+			reader = body
+		} else {
+			switch contentType {
+			case definition.MIMEJSON:
+				if err := json.NewEncoder(buf).Encode(r.body); err != nil {
+					return err
+				}
+			case definition.MIMEXML:
+				if err := xml.NewEncoder(buf).Encode(r.body); err != nil {
+					return err
+				}
+			default:
+				if _, err := buf.WriteString(fmt.Sprint()); err != nil {
+					return err
+				}
+			}
+		}
+	} else {
+		if len(r.files) > 0 {
+			parts := multipart.NewWriter(buf)
+			for k, v := range r.forms {
+				w, err := parts.CreateFormField(k)
+				if err != nil {
+					return err
+				}
+				if _, err = fmt.Fprint(w, v); err != nil {
+					return err
+				}
+			}
+			for k, v := range r.files {
+				w, err := parts.CreateFormFile(k, k)
+				if err != nil {
+					return err
+				}
+				if r, ok := v.(io.Reader); ok {
+					if _, err := io.Copy(w, r); err != nil {
+						return err
+					}
+				} else {
+					switch data := v.(type) {
+					case []byte:
+						if _, err := w.Write(data); err != nil {
+							return err
+						}
+					default:
+						if _, err = fmt.Fprint(w, v); err != nil {
+							return err
+						}
+					}
+				}
+			}
+			contentType = parts.FormDataContentType()
+		} else if len(r.forms) > 0 {
+			contentType = definition.MIMEURLEncoded
+			formVal := url.Values{}
+			for k, v := range r.forms {
+				formVal.Add(k, fmt.Sprint(v))
+			}
+			_, err := buf.WriteString(formVal.Encode())
+			if err != nil {
+				return err
+			}
+		}
+	}
+	req, err := http.NewRequest(r.method, path, reader)
+	if err != nil {
+		return err
+	}
+	for k, v := range r.headers {
+		req.Header.Add(k, fmt.Sprint(v))
+	}
+	req.Header.Set("Content-Type", contentType)
+	resp, err := http.DefaultClient.Do(req)
```

Comment:
every request will use a new client? i think use be add default client into NewRest?